flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/materializedtable/AlterMaterializedTableAsQueryOperation.java

```diff
@@ -0,0 +1,142 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations.materializedtable;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.api.internal.TableResultImpl;
+import org.apache.flink.table.api.internal.TableResultInternal;
+import org.apache.flink.table.catalog.CatalogMaterializedTable;
+import org.apache.flink.table.catalog.Column;
+import org.apache.flink.table.catalog.ObjectIdentifier;
+import org.apache.flink.table.catalog.ResolvedCatalogBaseTable;
+import org.apache.flink.table.catalog.ResolvedCatalogMaterializedTable;
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.catalog.TableChange;
+import org.apache.flink.table.operations.Operation;
+import org.apache.flink.table.operations.OperationUtils;
+import org.apache.flink.table.operations.QueryOperation;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.catalog.CatalogBaseTable.TableKind.MATERIALIZED_TABLE;
+
+/** Operation to describe an ALTER MATERIALIZED TABLE AS query operation. */
+@Internal
+public class AlterMaterializedTableAsQueryOperation extends AlterMaterializedTableOperation {
+
+    protected final ObjectIdentifier tableIdentifier;
+
+    private final QueryOperation queryOperation;
+
+    public AlterMaterializedTableAsQueryOperation(
+            ObjectIdentifier tableIdentifier, QueryOperation queryOperation) {
+        super(tableIdentifier);
+        this.tableIdentifier = tableIdentifier;
+        this.queryOperation = queryOperation;
+    }
+
+    public QueryOperation getQueryOperation() {
+        return queryOperation;
+    }
+
+    @Override
+    public TableResultInternal execute(Context ctx) {
+        ResolvedCatalogBaseTable<?> resolvedCatalogBaseTable =
+                ctx.getCatalogManager().getTableOrError(tableIdentifier).getResolvedTable();
+        if (MATERIALIZED_TABLE != resolvedCatalogBaseTable.getTableKind()) {
+            throw new ValidationException(
+                    String.format(
+                            "Table %s is not a materialized table, does not support materialized table related operation.",
+                            tableIdentifier));
+        }
+
+        ResolvedCatalogMaterializedTable oldResolvedMaterializedTable =
+                (ResolvedCatalogMaterializedTable) resolvedCatalogBaseTable;
+
+        // validate new schema and derived origin primary key and watermark spec
+        ResolvedSchema resolvedQuerySchema = queryOperation.getResolvedSchema();
```

Comment:
After deep consideration, I believe we should move the logic for validating the query, deriving the schema, and validating the added columns to the SqlAlterMaterializedTableAsQueryConverter. This way, we can obtain the corresponding definition query, TableChange, and CatalogMaterializedTable objects to serve as members of the AlterMaterializedTableAsQueryOperation.